torch_geometric/nn/models/pg_explainer.py

```diff
@@ -0,0 +1,270 @@
+            information based on the number of
+            :class:`~torch_geometric.nn.conv.message_passing.MessagePassing`
+            layers inside :obj:`model`. (default: :obj:`None`)
+        task (str): Denotes the type of task that needs explanation. Valid
+            inputs are :obj:`"node"` (for node classification) and
+            :obj:`"graph"` (for graph classification). (default: :obj:`"node"`)
+        return_type (str, optional): Denotes the type of output from
+            :obj:`model`. Valid inputs are :obj:`"log_prob"` (the model
+            returns the logarithm of probabilities), :obj:`"prob"` (the
+            model returns probabilities), :obj:`"raw"` (the model returns raw
+            scores) and :obj:`"regression"` (the model returns scalars).
+            (default: :obj:`"log_prob"`)
+        log (bool, optional): If set to :obj:`False`, will not log any learning
+            progress. (default: :obj:`True`)
+        **kwargs (optional): Additional hyper-parameters to override default
+            settings in :attr:`~torch_geometric.nn.models.GNNExplainer.coeffs`.
+    """
+
+    coeffs = {
+        'edge_size': 0.05,
+        'edge_ent': 1.0,
+        'temp': [5.0, 2.0],
+        'bias': 0
+    }
+
+    def __init__(self, model, out_channels: int, epochs: int = 30,
+                 lr: float = 0.003, num_hops: Optional[int] = None,
+                 task: str = 'node', return_type: str = 'log_prob',
+                 log: bool = True, **kwargs):
+        super().__init__(model, lr, epochs, num_hops, return_type, log)
+        assert task in ['node', 'graph']
+
+        self.out_channels = out_channels
+        self.task = task
+        self.coeffs.update(kwargs)
+
+        self.exp_in_channels = 2 * out_channels if (
+            task == 'graph') else 3 * out_channels
+        self.explainer_model = nn.Sequential(
+            nn.Linear(self.exp_in_channels, 64),
+            nn.ReLU(),
+            nn.Linear(64, 1),
+        )
+
+    def _get_temp(self, e: int) -> float:
+        temp = self.coeffs['temp']
+        return temp[0] * ((temp[1] / temp[0])**(e / self.epochs))
+
+    def _create_explainer_input(self, edge_index, x, node_idx=None) -> Tensor:
+        rows, cols = edge_index
+        x_j, x_i = x[rows], x[cols]
+        if self.task == 'node':
+            x_node = x[node_idx].repeat(rows.size(0), 1)
+            return torch.cat([x_i, x_j, x_node], 1)
+        else:
+            return torch.cat([x_i, x_j], 1)
+
+    def _compute_edge_mask(self, edge_weight, temperature=1.0, bias=0.0,
+                           training=True):
+
+        if training:  # noise is added to edge_weight.
+            bias += 0.0001
+            eps = (bias -
+                   (1 - bias)) * torch.rand(edge_weight.size()) + (1 - bias)
+            eps = eps.to(edge_weight.device)
+            return (eps.log() -
+                    (1 - eps).log() + edge_weight).squeeze() / temperature
```

Comment:
I think the squeeze() here and on line 107 need to be limited to certain dims. I hit a 2 node / 1 edge subgraph while exploring this code, and the squeeze reduced the edge_weight to a scalar tensor, causing IndexErrors later.  I tried it with squeeze(dim=1) in these two places and that worked, but I'm not familiar enough with the pg_explainer code to know if this edge_weight can ever have more than 2 dimensions.

Reply:
This may also be needed in the squeeze on line 143.

Reply:
I added a test for a 2 node 1 edge graph, seems to work fine
```
    # test with 2 node 1 edge graph:
    explainer = PGExplainer(model, out_channels=z.shape[1], task='node',
                            log=False, return_type=return_type)
    explainer.train_explainer(x[:2], z[:2], edge_index[:, :1],
                              node_idxs=torch.arange(0, 1))
    assert_mask_clear(model)
    mask = explainer.explain(x[:2], z[:2], edge_index[:, :1], node_idx=1)
    assert mask.shape[0] == 1
```
Does this test cover the scenario you mentioned?

Reply:
The test works even when the `dim=1` argument is not used in `squeeze`.

Reply:
I got the same error when using the second node as a training instance, instead of the first. 

`explainer.train_explainer(x[:2], z[:2], edge_index[:, :1],
                              node_idxs=torch.arange(1, 2))`

Reply:
I think the problem still exists when nodes with only self-connections are considered.

Reply:
Thanks @RBendias  and @SensorLock for bringing this up. Both issues 1. node with 1 incoming edge and 2. node with only self loop are the same and fixed by setting `dim=1` as @SensorLock  suggested.